pkg/request/pool.go

```diff
@@ -228,27 +201,67 @@ func (rp *Pool) RemoveRequests(requests ...string) error {
 	return nil
 }
 
-// Close removes all the requests, stops all the timeout timers.
+// Close closes the pool
 func (rp *Pool) Close() {
 	atomic.StoreUint32(&rp.closed, 1)
-	rp.Clear()
+	// TODO need to remove all requests?
+}
+
+func (rp *Pool) isClosed() bool {
+	return atomic.LoadUint32(&rp.closed) == 1
 }
 
-// StopTimers stops all the timeout timers attached to the pending requests, and marks the pool as "stopped".
-// This which prevents submission of new requests, and renewal of timeouts by timer go-routines that where running
-// at the time of the call to StopTimers().
-func (rp *Pool) StopTimers() {
+// Halt stops the callbacks of the first and second strikes.
+func (rp *Pool) Halt() {
 	atomic.StoreUint32(&rp.stopped, 1)
+	rp.pending.Stop()
+}
 
-	rp.logger.Debugf("Stopped all timers")
+func (rp *Pool) isStopped() bool {
+	return atomic.LoadUint32(&rp.stopped) == 1
 }
 
-// RestartTimers restarts all the timeout timers attached to the pending requests, as RequestForwardTimeout, and re-allows
-// submission of new requests.
-func (rp *Pool) RestartTimers() {
-	rp.lock.Lock()
-	defer rp.lock.Unlock()
+// Restart restarts the pool.
+// When batching is set to true the pool is expected to respond to NextRequests.
+func (rp *Pool) Restart(batching bool) {
+
+	defer atomic.StoreUint32(&rp.stopped, 0)
+
+	rp.Halt()
+
+	isBatching := rp.isBatchingEnabled()
+
+	rp.setBatching(batching)
+
+	if isBatching {
```

Comment:
can we check the two true conditions and the two false conditions first and then return early? 

Then we only will need to check `batching` alone.